### ISSUE TITLE
Is there a way to print latest AUR comments before starting the build?

### ISSUE THREAD

**Issue by Rogach** [polygamma/aurman#204]
Is it possible to instruct aurman to print latest comments from AUR before proceeding to package building, in order to see if there was some controversy regarding the package?

yaourt seems to do this: https://github.com/archlinuxfr/yaourt/issues/358

(I thought about it after reading this: https://sensorstechforum.com/arch-linux-aur-repository-found-contain-malware/)

**Comment by polygamma** [comment]
possible: yes, but i am not going to implement it. i would also not merge a pr implementing that.

**Comment by Rogach** [comment]
Could you please elaborate why you are against it? It seems to be improving the security slightly.

**Comment by polygamma** [comment]
ugly parsing of the aur pages is needed, there is no api providing the comments as far as i know. if there is one: forget what i said, in that case i'd implement it

**Comment by Rogach** [comment]
I feel your pain, parsing html pages gets horrible really fast. No, AUR's API does not export the comments:
https://github.com/lfos/aurweb/blob/master/web/lib/aurjson.class.php#L272

Thanks for the clarification!

**Comment by AladW** [comment]
`lynx --dump`

**Comment by AladW** [comment]
Because then there's no HTML parsing involved, you just take the relevant sections.
```
[61]Pinned Comments
(...)
[68]All comments
```

**Comment by AladW** [comment]
Compare that one-liner to yaourt's attempts, which unsurprisingly broke on AUR changes: https://github.com/archlinuxfr/yaourt/blob/master/src/lib/aur.sh#L97-L153

**Comment by Rogach** [comment]
Okay, so

```bash
$ lynx --dump https://aur.archlinux.org/packages/aurutils | sed -n '/Latest Comments/,/All comments/p' | head -n -2 | tail -n +3
```

indeed dumps the comments and looks something like this: https://gist.github.com/Rogach/07f1ed88d6b71c1a76c5d8efae99d129

But I agree with the author, that looks quite hacky and adds lynx to dependencies, going the AUR API way feels way better.

**Comment by AladW** [comment]
I don't see what's hacky about it (nor did the author reply yet), since it uses an actual webbrowser to access the contents, and those section names are unlikely to change. 

About the dependency, it would be optional since displaying AUR comments in a helper is not mandatory, just an (arguably) nice thing to have.

**Comment by Rogach** [comment]
True, I can't speak for the author. But I feel that there is a difference between "unlikely to change" (which sadly still sometimes changes) and being exactly specified in API (which even supports backward compatibility with older clients via `v` parameter: https://wiki.archlinux.org/index.php/Aurweb_RPC_interface).

**Comment by AladW** [comment]
`aurweb` has releases too, and these are infrequent. If a section name changes in such a release (the last change was the introduction of pinned comments), the required change is trivial, unlike when doing actual parsing by hand.

Anyway you can propose an extension for the AUR API on aur-dev. Until then, using a browser is the best you can do.

**Comment by eli-schwartz** [comment]
I somehow don't feel that subprocessing out to lynx in order to parse comment blocks by guessing based on trigger words is very reasonable. Yaourt sucks for using grep and sed, but the solution isn't to do even more of that.

Comments on the AUR have been tagged with the HTML class "article-content" since *2012*. Parsers which *don't use that* will obviously break on insignificant changes up to and including the addition of whitespace...

**Comment by eli-schwartz** [comment]
Addendum: before that, there was effectively no HTML tagging at all.

As for using HTML classes, the first rule of programmatically *anything* in HTML, is to try using XPath to select nodes semantically. Basically every programming language has libraries to do it, most standardized -- for Python, that means the highly popular python-lxml bindings to libxml2/libxslt. For bash, that typically means xmlstarlet.

The second rule is if you cannot, then give up and go home and quit the internet instead of trying to use regex.

(Yaourt chose option three.)

**Comment by AladW** [comment]
Hm. Care to post some `xmlstarlet` commandline that doesn't spew out 2 pages of errors?

I guess with Python you'd use this: https://lxml.de/elementsoup.html

**Comment by polygamma** [comment]
Sorry, I am not interested in implementing this without proper API usage. One may simply open a browser and open the AUR page of the relevant package to see the comments. The feature is nice to have, and I see using lynx for this as a hacky solution, too

**Comment by AladW** [comment]
I'm reminded of https://github.com/mortbauer/aur-comment-fetcher/blob/master/aur-comment-fetch, which uses both "trigger words" and beautiful soup. I'll stick to my perfectly functional one-liner. :slightly_smiling_face: (note to self: `elinks` has more `dump` options)
